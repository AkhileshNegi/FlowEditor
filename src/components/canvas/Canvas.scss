@import '~/variables.scss';

.canvas {
    position: relative;
    height: 10000px;
    width: 100%;
}

.dragSelection {
    position: absolute;
    background: $selection_square;
    border: 0px solid $selection_square - #222;
    opacity: 0.2;
    pointer-events: none;
    z-index: $z_selection;
}
