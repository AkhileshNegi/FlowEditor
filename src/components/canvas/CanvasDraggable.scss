@import '~/variables.scss';

.draggable {
    position: absolute;

    -webkit-touch-callout: none;
    -webkit-user-select: none;
    -khtml-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
}

.selected {
    z-index: $z_dragging;
}
