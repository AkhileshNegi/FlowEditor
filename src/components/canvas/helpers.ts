import { FlowPosition } from '~/flowTypes';
import { CanvasPositions } from '~/store/editor';

export const collides = (a: FlowPosition, b: FlowPosition) => {
    // don't bother with collision if we don't have full dimensions
    /* istanbul ignore next */
    if (!a.bottom || !b.bottom) {
        return false;
    }

    return !(b.left > a.right || b.right < a.left || b.top > a.bottom || b.bottom < a.top);
};

export const getCollisions = (
    positions: CanvasPositions,
    box: FlowPosition
): { [uuid: string]: FlowPosition } => {
    const collisions = {};
    for (const nodeUUID of Object.keys(positions)) {
        const position = positions[nodeUUID];
        if (collides(box, position)) {
            collisions[nodeUUID] = position;
        }
    }
    return collisions;
};
