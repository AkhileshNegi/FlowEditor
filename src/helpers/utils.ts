import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition } from '../flowTypes';

const request = require('sync-request');

const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase().trim(), SNAKED_CHARS, '_', 'all');
}

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export function getDisplayName(HOCName: string, Component: ComponentClass | SFC): string {
    const ComponentDisplayName = Component.displayName || Component.name || 'Component';
    return `${HOCName}(${ComponentDisplayName})`;
}

function getFlow(name: string): FlowDefinition {
    const definition = request('GET', 'base/test_flows/' + name + '.json').getBody();
    return JSON.parse(definition) as FlowDefinition;
}

export function getFavorites(): FlowDefinition {
    return getFlow('favorites');
}

export function getTest(): FlowDefinition {
    return getFlow('a4f64f1b-85bc-477e-b706-de313a022979');
}

export function dump(object: any) {
    console.log(JSON.stringify(object, null, 1));
}

export const DATA_SPEC_ATTRIBUTE_NAME: string = 'data-spec';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} matching DOM components
 */
export function getSpecWrapper(
    componentWrapper: ReactWrapper | ShallowWrapper,
    specName: string
): ShallowWrapper | ReactWrapper {
    return componentWrapper.find(`[${DATA_SPEC_ATTRIBUTE_NAME}="${specName}"]`);
}

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export function validUUID(uuid: string): boolean {
    const regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
    return regex.test(uuid);
}
