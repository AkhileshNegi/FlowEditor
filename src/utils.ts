const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {BoolMap} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase(), SNAKED_CHARS, '_', 'all');
}
