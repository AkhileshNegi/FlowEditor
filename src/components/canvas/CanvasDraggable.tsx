import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { Dimensions, FlowPosition } from '~/flowTypes';
import { newPosition } from '~/store/helpers';

import * as styles from './CanvasDraggable.scss';

export interface CanvasDraggableProps {
    position: FlowPosition;
    uuid: string;
    ele: (selected: boolean) => JSX.Element;

    selected?: boolean;

    onAnimated?: (uuid: string) => void;
    updateDimensions?: (uuid: string, position: Dimensions) => void;
    onDragStart?: (uuid: string, clickedPosition: FlowPosition) => void;
    onDragStop?: () => void;
}

export class CanvasDraggable extends React.PureComponent<CanvasDraggableProps, {}> {
    private ele: HTMLDivElement;

    constructor(props: CanvasDraggableProps) {
        super(props);
        bindCallbacks(this, {
            include: [/^handle/, 'ref']
        });
    }

    private ref(ref: HTMLDivElement): any {
        return (this.ele = ref);
    }

    public componentDidMount(): void {
        if (this.ele) {
            if (this.ele.clientWidth && this.ele.clientHeight) {
                this.props.updateDimensions(this.props.uuid, {
                    width: this.ele.clientWidth,
                    height: this.ele.clientHeight
                });
            }
        }
    }

    public componentDidUpdate(prevProps: CanvasDraggableProps): void {
        if (this.ele) {
            if (this.ele.clientWidth && this.ele.clientHeight) {
                this.props.updateDimensions(this.props.uuid, {
                    width: this.ele.clientWidth,
                    height: this.ele.clientHeight
                });
            }
        }
    }

    public render(): JSX.Element {
        const classes = [styles.draggable];

        if (this.props.selected) {
            classes.push(styles.selected);
        }

        const handleAnimated = () => {
            if (this.props.onAnimated) {
                this.props.onAnimated(this.props.uuid);
            }
        };

        return (
            <div
                onTransitionEnd={handleAnimated}
                ref={this.ref}
                className={classes.join(' ')}
                style={{ left: this.props.position.left, top: this.props.position.top }}
                onMouseDown={(event: React.MouseEvent<HTMLDivElement>) => {
                    this.props.onDragStart(
                        this.props.uuid,
                        newPosition(
                            event.pageX - this.props.position.left,
                            event.pageY - this.props.position.top
                        )
                    );
                }}
                onMouseUp={(event: React.MouseEvent<HTMLDivElement>) => {
                    this.props.onDragStop();
                }}
            >
                {this.props.ele(this.props.selected)}
            </div>
        );
    }
}
