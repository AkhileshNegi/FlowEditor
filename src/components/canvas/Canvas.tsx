import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { CanvasDraggable, CanvasDraggableProps } from '~/components/canvas/CanvasDraggable';
import { DRAG_THRESHOLD } from '~/components/flow/Flow';
import { FlowPosition } from '~/flowTypes';
import { CanvasPositions, DragSelection } from '~/store/editor';
import { MergeEditorState } from '~/store/thunks';
import { createUUID } from '~/utils';

import * as styles from './Canvas.scss';

interface CanvasProps {
    dragActive: boolean;
    draggables: CanvasDraggableProps[];
    getCurrentPosition: (uuid: string) => FlowPosition;
    onUpdateDragPositions: (
        delta: FlowPosition,
        originalPositions: CanvasPositions,
        snap: boolean
    ) => void;
    onCheckCollisions: (box: FlowPosition) => CanvasPositions;
    canvasSelections: CanvasPositions;
    mergeEditorState: MergeEditorState;
}

interface CanvasState {
    dragDownPosition: FlowPosition;
    dragUUID: string;
    dragGroup: boolean;
    dragSelection: DragSelection;
    uuid: string;
}

export class Canvas extends React.Component<CanvasProps, CanvasState> {
    private ele: HTMLDivElement;
    private parentOffset: FlowPosition;

    constructor(props: CanvasProps) {
        super(props);

        this.state = {
            dragDownPosition: null,
            dragUUID: null,
            dragGroup: false,
            dragSelection: null,
            uuid: createUUID()
        };

        bindCallbacks(this, {
            include: [/^handle/, /^render/]
        });
    }

    public componentDidMount(): void {
        let offset = { left: 0, top: 0 };
        /* istanbul ignore next */
        if (this.ele) {
            offset = this.ele.getBoundingClientRect();
        }
        this.parentOffset = { left: offset.left, top: offset.top + window.scrollY };
    }

    public renderSelectionBox(): JSX.Element {
        const drag = this.state.dragSelection;

        if (drag) {
            const left = Math.min(drag.startX, drag.currentX);
            const top = Math.min(drag.startY, drag.currentY);
            const width = Math.max(drag.startX, drag.currentX) - left;
            const height = Math.max(drag.startY, drag.currentY) - top;

            if (this.state.dragSelection && this.state.dragSelection.startX) {
                return (
                    <div className={styles.dragSelection} style={{ left, top, width, height }} />
                );
            }
        }

        return null;
    }

    private isClickOnCanvas(event: React.MouseEvent<HTMLDivElement>): boolean {
        return (event.target as any).id === this.state.uuid;
    }

    private handleMouseDown(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            this.setState({
                dragSelection: {
                    startX: event.pageX - this.parentOffset.left,
                    startY: event.pageY - this.parentOffset.top,
                    currentX: event.pageX - this.parentOffset.left,
                    currentY: event.pageY - this.parentOffset.top
                }
            });
        }
    }

    private handleMouseMove(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.state.dragSelection && this.state.dragSelection.startX) {
            const drag = this.state.dragSelection;
            const left = Math.min(drag.startX, drag.currentX);
            const top = Math.min(drag.startY, drag.currentY);
            const right = Math.max(drag.startX, drag.currentX);
            const bottom = Math.max(drag.startY, drag.currentY);

            const selected = this.props.onCheckCollisions({ left, top, right, bottom });
            this.setState({
                dragSelection: {
                    startX: drag.startX,
                    startY: drag.startY,
                    currentX: event.pageX - this.parentOffset.left,
                    currentY: event.pageY - this.parentOffset.top
                }
            });

            this.props.mergeEditorState({ canvasSelections: selected });
        }

        if (this.state.dragUUID) {
            this.updatePositions(event, false);
        }
    }

    private updatePositions(event: React.MouseEvent<HTMLDivElement>, snap: boolean) {
        const { dragUUID } = this.state;

        const startPosition = this.props.dragActive
            ? this.props.canvasSelections[dragUUID]
            : this.props.getCurrentPosition(dragUUID);

        const xd =
            event.pageX -
            this.parentOffset.left -
            this.state.dragDownPosition.left -
            startPosition.left;

        const yd =
            event.pageY -
            this.parentOffset.top -
            this.state.dragDownPosition.top -
            startPosition.top;

        if (this.props.dragActive) {
            this.props.onUpdateDragPositions(
                { left: xd, top: yd },
                this.props.canvasSelections,
                snap
            );
        } else {
            if (Math.abs(xd) + Math.abs(yd) > DRAG_THRESHOLD) {
                let selected = this.props.canvasSelections;
                if (!(this.state.dragUUID in selected)) {
                    selected = { [dragUUID]: this.props.getCurrentPosition(dragUUID) };
                }

                this.props.mergeEditorState({
                    dragActive: true,
                    canvasSelections: selected
                });
            }
        }
    }

    private handleMouseUp(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.state.dragUUID) {
            this.updatePositions(event, true);

            this.setState({
                dragDownPosition: null,
                dragSelection: null,
                dragUUID: null
            });

            this.props.mergeEditorState({
                canvasSelections: {},
                dragActive: false
            });
        }

        if (this.state.dragSelection && this.state.dragSelection.startX) {
            this.setState({
                dragSelection: {
                    startX: null,
                    startY: null,
                    currentX: null,
                    currentY: null
                }
            });
        }
    }

    private handleDragStart(uuid: string, position: FlowPosition): void {
        this.setState({
            dragUUID: uuid,
            dragDownPosition: {
                left: position.left - this.parentOffset.left,
                top: position.top - this.parentOffset.top
            }
        });
    }

    private handleDragStop(): void {
        this.setState({
            dragUUID: null,
            dragDownPosition: null,
            dragSelection: null
        });

        this.props.mergeEditorState({
            dragActive: false
        });
    }

    public render(): JSX.Element {
        return (
            <div
                id={this.state.uuid}
                ref={(ele: HTMLDivElement) => {
                    this.ele = ele;
                }}
                className={styles.canvas}
                onMouseDown={this.handleMouseDown}
                onMouseMove={this.handleMouseMove}
                onMouseUp={this.handleMouseUp}
            >
                {this.props.draggables.map((draggable: CanvasDraggableProps) => (
                    <CanvasDraggable
                        selected={!!this.props.canvasSelections[draggable.uuid]}
                        key={draggable.ele.key}
                        {...draggable}
                        onDragStart={this.handleDragStart}
                        onDragStop={this.handleDragStop}
                    />
                ))}
                {this.props.children}
                {this.renderSelectionBox()}
            </div>
        );
    }
}
