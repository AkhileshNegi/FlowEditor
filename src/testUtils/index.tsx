// TODO: Remove use of Function
// tslint:disable:ban-types
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { object } from 'prop-types';
import * as React from 'react';
import * as config from '../../__test__/config';
import { ConfigProviderContext, endpointsPT, flowPT, languagesPT } from '../config';
import { FlowDefinition, FlowEditorConfig } from '../flowTypes';
import { AppState, createStore, initialState } from '../store';
import { getBaseLanguage } from '../utils';

export interface Resp {
    results: Array<{ [key: string]: any }>;
    total?: number;
}

export interface QueryString {
    [key: string]: string;
}

const initialContext: ConfigProviderContext = {
    endpoints: (config as FlowEditorConfig).endpoints,
    languages: (config as FlowEditorConfig).languages,
    flow: (config as FlowEditorConfig).flow
};

const initialTestState: AppState = {
    ...initialState,
    ...{
        flowContext: {
            ...initialState.flowContext,
            definition: require('../../__test__/flows/colors.json') as FlowDefinition
        }
    },
    ...{
        flowEditor: {
            ...initialState.flowEditor,
            editorUI: {
                ...initialState.flowEditor.editorUI,
                language: getBaseLanguage((config as FlowEditorConfig).languages)
            }
        }
    }
};

/**
 * Compose setup method for component tests
 */
export const createSetup = <P extends {}>(
    Component: React.ComponentClass | React.SFC,
    baseProps: P = {} as any,
    baseDuxState: AppState | Partial<AppState> = initialState,
    baseContext: ConfigProviderContext = initialContext,
    connectedChildren: boolean = false
) => (
    shallowRender: boolean = true,
    propOverrides: P | Partial<P> = {},
    duxStateOverrides = {} as AppState,
    contextOverrides = {} as ConfigProviderContext | Partial<ConfigProviderContext>,
    childContextTypeOverrides: { [key: string]: Function } = {}
) => {
    // Waiting on https://github.com/Microsoft/TypeScript/pull/1328
    const props = Object.assign({}, baseProps, propOverrides);
    const context = connectedChildren
        ? Object.assign({}, baseContext, {
              store: createStore(Object.assign({}, baseDuxState, duxStateOverrides))
          })
        : baseContext;
    // tslint:disable-next-line:ban-types
    const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
        <Component {...props} />,
        {
            context,
            childContextTypes: Object.assign(
                {},
                {
                    store: object,
                    endpoints: endpointsPT,
                    languages: languagesPT,
                    flow: flowPT
                },
                childContextTypeOverrides
            )
        }
    );

    return {
        wrapper,
        props,
        context,
        instance: wrapper.instance()
    };
};
export const createSpy = (obj: Object | React.ComponentClass) => (instanceMethod: string) =>
    jest.spyOn((obj as React.ComponentClass).prototype || obj, instanceMethod as any);

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () => new Promise(resolve => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance[]) =>
    spies.forEach(spy => spy.mockRestore());

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
    specName: string,
    attributeName: string = 'data-spec'
): any => {
    return componentWrapper.find(`[${attributeName}="${specName}"]`);
};

const DEBUG = false;

export const timeStart = (name: string) => {
    if (DEBUG) {
        console.time(name);
    }
};

export const timeEnd = (name: string) => {
    if (DEBUG) {
        console.timeEnd(name);
    }
};
