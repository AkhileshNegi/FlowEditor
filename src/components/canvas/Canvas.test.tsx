import { Canvas, CANVAS_PADDING, CanvasProps } from '~/components/canvas/Canvas';
import { CanvasDraggableProps } from '~/components/canvas/CanvasDraggable';
import { composeComponentTestUtils, mock } from '~/testUtils';
import * as utils from '~/utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const ele = (selected: boolean): JSX.Element => null;

const baseProps: CanvasProps = {
    uuid: utils.createUUID(),
    dragActive: false,
    onDragged: jest.fn(),
    onUpdateDragPositions: jest.fn(),
    mergeEditorState: jest.fn(),
    draggables: []
};

const { setup } = composeComponentTestUtils(Canvas, baseProps);

describe(Canvas.name, () => {
    it('render default', () => {
        const { wrapper } = setup();
        expect(wrapper).toMatchSnapshot();
    });

    it('initializes the height to 1000', () => {
        const { instance } = setup();
        expect(instance.state.height).toEqual(1000);
    });

    it('initializes the height to the lowest draggable', () => {
        const lowest: CanvasDraggableProps = {
            ele,
            uuid: utils.createUUID(),
            position: { top: 1200, left: 100, bottom: 1290, right: 300 }
        };

        const { instance } = setup(true, { $set: { draggables: [lowest] } });
        expect(instance.state.height).toEqual(1290 + CANVAS_PADDING);
    });

    it('adjust the height when updating dimensions', () => {
        const lowest: CanvasDraggableProps = {
            ele,
            uuid: utils.createUUID(),
            position: { top: 1200, left: 100 }
        };
        const { instance } = setup(true, { $set: { draggables: [lowest] } });
        expect(instance.state.height).toEqual(1000);

        instance.handleUpdateDimensions(lowest.uuid, { width: 200, height: 300 });
        expect(instance.state.height).toBe(lowest.position.top + 300 + CANVAS_PADDING);
    });
});
