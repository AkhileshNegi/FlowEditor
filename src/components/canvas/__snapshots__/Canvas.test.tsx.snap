// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas render default 1`] = `
<div
  className="canvasContainer"
>
  <div
    className="canvas"
    id="b1f332f3-bdd3-4891-aec5-1843a712dbf1"
    onMouseDown={[Function]}
    onMouseMove={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "height": 1000,
      }
    }
  />
</div>
`;
